### PR TITLE
Adds Docker provider support

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

Adds Docker provider support.

### Motivation

To have the Docker provider.

### More

- [x] Added/updated tests
- [ ] Added/updated documentation

Co-authored-by: Julien Salleyron <julien@containo.us>

### CODE REVIEW

File: provider/docker/config_test.go [hunk 1]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with two different middleware with same name",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with two different middlewares with same name",
```

File: provider/docker/config_test.go [hunk 2]
```diff
@@ -0,0 +1,2138 @@
...
+						MaxConn: &config.MaxConn{
+							Amount:        42,
+							ExtractorFunc: "request.host",
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with two identical middleware",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with two identical middlewares",
```

File: provider/docker/config_test.go [hunk 3]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "three containers with different middleware with same name",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "three containers with different middlewares with same name",
```

File: provider/docker/config_test.go [hunk 4]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with two different router with same name",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with two different routers with same name",
```

File: provider/docker/config_test.go [hunk 5]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "three containers with different router with same name",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "three containers with different routers with same name",
```

File: provider/configuration.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+			middlewares[middlewareName] = append(middlewares[middlewareName], root)
+			if !AddMiddleware(configuration, middlewareName, middleware) {
+				middlewaresToDelete[middlewareName] = struct{}{}
+			}
+		}
+	}
+
+	for serviceName := range servicesToDelete {
+		logger.WithField(log.ServiceName, serviceName).
+			Errorf("Service defined multiple times with different configuration in %v", services[serviceName])
```
Comment (jbdoumenjou) [inline provider/configuration.go]:
```suggestion
			Errorf("Service defined multiple times with different configurations in %v", services[serviceName])
```

File: provider/configuration.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		delete(configuration.Services, serviceName)
+	}
+
+	for routerName := range routersToDelete {
+		logger.WithField(log.RouterName, routerName).
+			Errorf("Router defined multiple times with different configuration in %v", routers[routerName])
```
Comment (jbdoumenjou) [inline provider/configuration.go]:
```suggestion
			Errorf("Router defined multiple times with different configurations in %v", routers[routerName])
```

File: provider/configuration.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		delete(configuration.Routers, routerName)
+	}
+
+	for middlewareName := range middlewaresToDelete {
+		logger.WithField(log.MiddlewareName, middlewareName).
+			Errorf("Middleware defined multiple times with different configuration in %v", middlewares[middlewareName])
```
Comment (jbdoumenjou) [inline provider/configuration.go]:
```suggestion
			Errorf("Middleware defined multiple times with different configurations in %v", middlewares[middlewareName])
```

File: provider/docker/config.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	"github.com/containous/traefik/provider/label"
+	"github.com/docker/go-connections/nat"
+)
+
+func (p *Provider) buildConfiguration(ctx context.Context, containersInspected []dockerData) *config.Configuration {
+	configurations := make(map[string]*config.Configuration)
+
+	for _, container := range containersInspected {
+		containerName := getServiceName(container) + "-" + container.ID
+
```
Comment (jbdoumenjou) [inline provider/docker/config.go]:
```suggestion
```

File: provider/docker/config.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *Provider) buildRouterConfiguration(ctx context.Context, container dockerData, configuration *config.Configuration) {
+	logger := log.FromContext(ctx)
+
```
Comment (jbdoumenjou) [inline provider/docker/config.go]:
```suggestion
```

File: provider/docker/config.go [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+	for routerName, router := range configuration.Routers {
+		if router.Rule == "" {
+			router.Rule = "Host:" + getSubDomain(serviceName) + "." + container.ExtraConf.Domain
+		}
+
+		if router.Service == "" {
+			if len(configuration.Services) > 1 {
+				delete(configuration.Routers, routerName)
+				logger.WithField(log.RouterName, routerName).
+					Error("could not define the service name for the router: too many services")
```
Comment (jbdoumenjou) [inline provider/docker/config.go]:
```suggestion
					Error("Could not define the service name for the router: too many services")
```

File: provider/docker/config_test.go [hunk 12]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with same service name and different LB method",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with same service name and different LB methods",
```

File: provider/docker/config_test.go [hunk 13]
```diff
@@ -0,0 +1,2138 @@
...
+						Service: "Service1",
+						Rule:    "Host:Test.traefik.wtf",
+					},
+				},
+				Middlewares: map[string]*config.Middleware{},
+				Services:    map[string]*config.Service{},
+			},
+		},
+		{
+			desc: "three containers with same service name and different LB method",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "three containers with same service name and different LB methods",
```

File: provider/docker/config_test.go [hunk 14]
```diff
@@ -0,0 +1,2138 @@
...
+						Service: "Service1",
+						Rule:    "Host:Test.traefik.wtf",
+					},
+				},
+				Middlewares: map[string]*config.Middleware{},
+				Services:    map[string]*config.Service{},
+			},
+		},
+		{
+			desc: "two containers with same service name and same LB method",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with same service name and same LB methods",
```

File: provider/docker/config_test.go [hunk 15]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with two identical router",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with two identical routers",
```

File: provider/docker/config_test.go [hunk 16]
```diff
@@ -0,0 +1,2138 @@
...
+							},
+							Method:         "wrr",
+							PassHostHeader: true,
+						},
+					},
+				},
+			},
+		},
+		{
+			desc: "two containers with two identical router rule and different service name",
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]:
```suggestion
			desc: "two containers with two identical router rules and different service names",
```

File: provider/docker/config_test.go [hunk 17]
```diff
@@ -0,0 +1,2138 @@
...
+		{
+			desc: "two networks, no network label, mode host, use provider network",
+			container: containerJSON(
+				networkMode("host"),
+				withNetwork("testnet", ipv4("10.11.12.13")),
+				withNetwork("webnet", ipv4("10.11.12.14")),
+			),
+			expected: "10.11.12.14",
+		},
+		{
```
Comment (jbdoumenjou) [inline provider/docker/config_test.go]: Is it a duplicate of the test L1969?

Comment (ldez) [reply provider/docker/config_test.go]: nope

Comment (jbdoumenjou) [reply provider/docker/config_test.go]: I don't know if it is important but don't we have to write a different test description ?

File: config/middlewares.go [hunk 18]
```diff
@@ -199,6 +199,11 @@ type MaxConn struct {
 	ExtractorFunc string `json:"extractorFunc,omitempty"`
 }
 
+// DefaultsHook Default values for a MaxConn.
```
Comment (geraldcroes) [inline config/middlewares.go]: maybe you should add the DefaultsHook to rateLimit too

File: provider/docker/docker.go [hunk 19]
```diff
@@ -48,20 +49,20 @@ type Provider struct {
...
+	ID              string
 	ServiceName     string
 	Name            string
 	Labels          map[string]string // List of labels set to container or service
 	NetworkSettings networkSettings
 	Health          string
 	Node            *dockertypes.ContainerNode
-	SegmentLabels   map[string]string
-	SegmentName     string
+	ExtraConf       configuration
```
Comment (geraldcroes) [inline provider/docker/docker.go]: A comment might be useful here

File: provider/docker/docker.go [hunk 20]
```diff
@@ -268,9 +282,18 @@ func listContainers(ctx context.Context, dockerClient client.ContainerAPIClient)
...
+		}
+
+		extraConf, err := p.getConfiguration(dData)
+		if err != nil {
+			log.FromContext(ctx).Errorf("Skip container %s: %v", getServiceName(dData), err)
+			continue
 		}
+		dData.ExtraConf = extraConf
+
+		containersInspected = append(containersInspected, dData)
```
Comment (geraldcroes) [inline provider/docker/docker.go]: to be nitpicky: inspectedContainers

File: server/middleware/middlewares.go [hunk 21]
```diff
@@ -314,6 +314,10 @@ func (b *Builder) buildConstructor(ctx context.Context, middlewareName string, c
 		}
 	}
 
+	if middleware == nil {
```
Comment (geraldcroes) [inline server/middleware/middlewares.go]: There is no test for this use case

File: types/tls.go [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+				}
+			} else {
+				return nil, fmt.Errorf("tls cert is a file, but tls key is not")
+			}
+		} else {
+			if errKeyIsFile != nil {
+				cert, err = tls.X509KeyPair([]byte(clientTLS.Cert), []byte(clientTLS.Key))
+				if err != nil {
+					return nil, fmt.Errorf("failed to load TLS keypair: %v", err)
+
```
Comment (jbdoumenjou) [inline types/tls.go]:
```suggestion
```

File: types/tls.go [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+	if clientTLS == nil {
+		log.FromContext(ctx).Warnf("clientTLS is nil")
+		return nil, nil
+	}
+
+	caPool := x509.NewCertPool()
+	clientAuth := tls.NoClientCert
+	if clientTLS.CA != "" {
+		var ca []byte
+		if _, errCA := os.Stat(clientTLS.CA); errCA == nil {
```
Comment (jbdoumenjou) [inline types/tls.go]:
```suggestion

		if _, errCA := os.Stat(clientTLS.CA); errCA == nil {
```

File: types/tls.go [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+			if err != nil {
+				return nil, fmt.Errorf("failed to read CA. %s", err)
+			}
+		} else {
+			ca = []byte(clientTLS.CA)
+		}
+		if !caPool.AppendCertsFromPEM(ca) {
+			return nil, fmt.Errorf("failed to parse CA")
+		}
+		if clientTLS.CAOptional {
```
Comment (jbdoumenjou) [inline types/tls.go]:
```suggestion

		if clientTLS.CAOptional {
```

File: types/tls.go [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+			clientAuth = tls.VerifyClientCertIfGiven
+		} else {
+			clientAuth = tls.RequireAndVerifyClientCert
+		}
+	}
+
+	cert := tls.Certificate{}
```
Comment (jbdoumenjou) [inline types/tls.go]: Could be a var after L62 or simply removed and we can remove err L62 and use assignment for L65 and L74.

File: types/tls.go [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+	_, errKeyIsFile := os.Stat(clientTLS.Key)
```
Comment (jbdoumenjou) [inline types/tls.go]: Could be set after L62

File: server/middleware/middlewares.go (hunk 21 above)
Comment (ldez) [reply server/middleware/middlewares.go]: Tests added

**Pull request merged**